pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+package registry
+
+import (
+	"errors"
+	"math"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/btree"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/origin/pkg/oauth/apis/oauth"
+	oauthclient "github.com/openshift/origin/pkg/oauth/generated/internalclientset/typed/oauth/internalversion"
+)
+
+var ErrTimedout = errors.New("Token timed out")
+
+type tokenData struct {
+	name    string
+	client  string
+	seen    time.Time
+	timeout time.Time
+}
+
+type tokenDataRef struct {
+	name    string
+	timeout time.Time
+}
+
+func (a *tokenDataRef) Less(than btree.Item) bool {
+
+	if a.timeout.Equal(than.(*tokenDataRef).timeout) {
+		return a.name < than.(*tokenDataRef).name
+	}
+	return a.timeout.Before(than.(*tokenDataRef).timeout)
+}
+
+type TokenTimeoutUpdater struct {
+	oauthClient    oauthclient.OAuthClientInterface
+	clientTimeouts map[string]int32
+	tokens         oauthclient.OAuthAccessTokenInterface
+	tokenChannel   chan tokenData
+	data           map[string]tokenData
+	tree           *btree.BTree
+}
+
+func NewTokenTimeoutUpdater(tokens oauthclient.OAuthAccessTokenInterface, oauthClient oauthclient.OAuthClientInterface) *TokenTimeoutUpdater {
+	ttu := &TokenTimeoutUpdater{
+		oauthClient,
+		make(map[string]int32),
+		tokens,
+		make(chan tokenData, 100),
+		make(map[string]tokenData),
+		// FIXME: what is the right degree for the btree
+		btree.New(32),
+	}
+	go ttu.Start()
```

Comment:
You will need to figure out how to plumb this to a post start hook.

Reply:
Why ?

Reply:
Because this runs forever, and thus leaks the goroutine, especially during tests.